src/ResourceManager/DataLakeStore/Commands.DataLakeStore/Models/DataLakeStoreTraceLogger.cs

```diff
@@ -0,0 +1,161 @@
+﻿// 
+// Copyright (c) Microsoft and contributors.  All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//   http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// 
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// 
+
+using Microsoft.Azure.Commands.DataLakeStore.Properties;
+using Microsoft.Rest;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Management.Automation;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Azure.Commands.DataLakeStore.Models
+{
+    public enum LogLevel
+    {
+        Debug = 0,
+        Information = 1,
+        Error = 2,
+        None = 3
+    }
+
+    public class DataLakeStoreTraceLogger : IDisposable
+    {
+        private const string MessagePrefixFormat = "ADLDataTransfer - ActivityId: [{0}] {1:yyyy-MM-dd HH:mm:ss.ffffff} T{2} {3}";
+
+        internal string LogFilePath { get; set; }
+
+        internal LogLevel LogLevel { get; set; }
+
+        internal FileStream TraceStream { get; set; }
+
+        internal TextWriterTraceListener TextListener { get; set; }
+
+        private bool PreviousAutoFlush { get; set; }
+
+        public readonly IServiceClientTracingInterceptor SdkTracingInterceptor;
+
+        public DataLakeStoreTraceLogger(Cmdlet commandToLog, string logFilePath = null, LogLevel logLevel = LogLevel.Information)
+        {
+            LogFilePath = logFilePath;
+            LogLevel = logLevel;
+            if (string.IsNullOrEmpty(LogFilePath))
+            {
+                LogFilePath = string.Format(@"{0}\ADLDataTransfer\ADLDataTransfer_{1}.log",
+                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
+                    DateTime.Now.ToString("MM-dd-yyyy.HH.mm"));
+            }
+            else if(Directory.Exists(LogFilePath)) // the user passed in a directory instead of a file
```

Comment:
Suggestion: You should consider usign IDataStore for file system operations - this allows you to mock the file system in tests.  You would need to augment it to allow returning a stream to support this entire scenario.